File: src/claw_mem/retrieval/query_cache.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Query Cache for claw-mem v2.5.0
LRU cache with TTL for search results
"""

import time
import hashlib
from typing import Any, Dict, List, Optional, Tuple
from collections import OrderedDict
from dataclasses import dataclass
import threading


@dataclass
class CacheEntry:
    """Cache entry with metadata"""
    results: List[Dict]
    timestamp: float
    access_count: int = 0


class QueryCache:
    """LRU Cache with TTL for query results

    Features:
    - LRU eviction
    - TTL expiration
    - Access frequency tracking
    - Thread-safe operations
    """

    def __init__(
        self,
        max_size: int = 1000,
        ttl_seconds: float = 300.0,
        min_access_count: int = 2
    ):
        """Initialize cache

        Args:
            max_size: Maximum number of entries
            ttl_seconds: Time-to-live in seconds
            min_access_count: Minimum accesses before frequency boost
        """
        self.max_size = max_size
        self.ttl_seconds = ttl_seconds
        self.min_access_count = min_access_count

        self._cache: OrderedDict[str, CacheEntry] = OrderedDict()
        self._lock = threading.RLock()

        # Stats
        self._hits = 0
        self._misses = 0

    def _make_key(self, query: str, top_k: int = 10) -> str:
        """Generate cache key from query"""
        key_data = f"{query.lower().strip()}:{top_k}"
        return hashlib.md5(key_data.encode()).hexdigest()

    def get(self, query: str, top_k: int = 10) -> Optional[List[Dict]]:
        """Get cached results

        Args:
            query: Search query
            top_k: Number of results

        Returns:
            Cached results or None
        """
        key = self._make_key(query, top_k)

        with self._lock:
            if key not in self._cache:
                self._misses += 1
                return None

            entry = self._cache[key]

            # Check TTL
            if time.time() - entry.timestamp > self.ttl_seconds:
                del self._cache[key]
                self._misses += 1
                return None

            # Update access order (move to end)
            self._cache.move_to_end(key)
            entry.access_count += 1

            self._hits += 1
            return entry.results

    def put(self, query: str, results: List[Dict], top_k: int = 10):
        """Cache results

        Args:
            query: Search query
            results: Search results
            top_k: Number of results
        """
        key = self._make_key(query, top_k)

        with self._lock:
            # Evict if needed
            while len(self._cache) >= self.max_size:
                self._cache.popitem(last=False)

            # Store
            self._cache[key] = CacheEntry(
                results=results,
                timestamp=time.time(),
                access_count=1
            )

    def invalidate(self, query: Optional[str] = None):
        """Invalidate cache

        Args:
            query: Specific query to invalidate, or None for all
        """
        with self._lock:
            if query is None:
                self._cache.clear()
            else:
                key = self._make_key(query)
                self._cache.pop(key, None)

    def get_stats(self) -> Dict[str, Any]:
        """Get cache statistics

        Returns:
            Statistics dictionary
        """
        with self._lock:
            total = self._hits + self._misses
            hit_rate = (self._hits / total * 100) if total > 0 else 0

            return {
                "size": len(self._cache),
                "max_size": self.max_size,
                "hits": self._hits,
                "misses": self._misses,
                "hit_rate": hit_rate,
                "ttl_seconds": self.ttl_seconds
            }

    def cleanup_expired(self):
        """Remove expired entries"""
        with self._lock:
            now = time.time()
            expired = [
                key for key, entry in self._cache.items()
                if now - entry.timestamp > self.ttl_seconds
            ]
            for key in expired:
                del self._cache[key]


# Global cache instance
_query_cache: Optional[QueryCache] = None


def get_query_cache(
    max_size: int = 1000,
    ttl_seconds: float = 300.0
) -> QueryCache:
    """Get global query cache instance"""
    global _query_cache

    if _query_cache is None:
        _query_cache = QueryCache(
            max_size=max_size,
            ttl_seconds=ttl_seconds
        )

    return _query_cache

File: src/retrieval/query_cache.ts
// claw-mem v5.0.0 — LRU Query Cache (TypeScript)
//
// LRU cache with TTL for search results.
// Pure TypeScript implementation with no external dependencies.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { createHash } from "crypto";

/** A single cache entry with metadata. */
interface CacheEntry {
  results: unknown[];
  timestamp: number;
  accessCount: number;
}

/**
 * LRU Cache with TTL for query results.
 *
 * Features:
 * - LRU eviction when max size is reached
 * - TTL expiration
 * - Access frequency tracking
 * - MD5-based cache keys
 */
export class QueryCache {
  private maxSize: number;
  private ttlSeconds: number;
  private minAccessCount: number;
  private cache: Map<string, CacheEntry> = new Map();
  private hits: number = 0;
  private misses: number = 0;

  constructor(maxSize: number = 1000, ttlSeconds: number = 300, minAccessCount: number = 2) {
    this.maxSize = maxSize;
    this.ttlSeconds = ttlSeconds;
    this.minAccessCount = minAccessCount;
  }

  /**
   * Generate a cache key from query and top_k.
   */
  private makeKey(query: string, topK: number = 10): string {
    const keyData = `${query.toLowerCase().trim()}:${topK}`;
    return createHash("md5").update(keyData).digest("hex");
  }

  /**
   * Get cached results for a query.
   *
   * @param query - Search query.
   * @param topK - Number of results.
   * @returns Cached results array, or undefined if cache miss / expired.
   */
  get(query: string, topK: number = 10): unknown[] | undefined {
    const key = this.makeKey(query, topK);
    const entry = this.cache.get(key);

    if (!entry) {
      this.misses++;
      return undefined;
    }

    // Check TTL
    if (Date.now() - entry.timestamp > this.ttlSeconds * 1000) {
      this.cache.delete(key);
      this.misses++;
      return undefined;
    }

    // LRU: delete and re-insert to move to end (most recently used)
    this.cache.delete(key);
    this.cache.set(key, entry);
    entry.accessCount++;

    this.hits++;
    return entry.results;
  }

  /**
   * Cache results for a query.
   *
   * @param query - Search query.
   * @param results - Search results to cache.
   * @param topK - Number of results.
   */
  put(query: string, results: unknown[], topK: number = 10): void {
    const key = this.makeKey(query, topK);

    // Evict LRU entries if at capacity
    while (this.cache.size >= this.maxSize) {
      const lruKey = this.cache.keys().next().value;
      if (lruKey !== undefined) {
        this.cache.delete(lruKey);
      } else {
        break;
      }
    }

    this.cache.set(key, {
      results,
      timestamp: Date.now(),
      accessCount: 1,
    });
  }

  /**
   * Invalidate cache entries.
   *
   * @param query - Specific query to invalidate, or undefined for all.
   */
  invalidate(query?: string): void {
    if (query === undefined) {
      this.cache.clear();
    } else {
      const key = this.makeKey(query);
      this.cache.delete(key);
    }
  }

  /**
   * Remove all expired entries.
   */
  cleanupExpired(): void {
    const now = Date.now();
    for (const [key, entry] of this.cache.entries()) {
      if (now - entry.timestamp > this.ttlSeconds * 1000) {
        this.cache.delete(key);
      }
    }
  }

  /**
   * Get cache statistics.
   */
  getStats(): Record<string, unknown> {
    const total = this.hits + this.misses;
    const hitRate = total > 0 ? (this.hits / total) * 100 : 0;

    return {
      size: this.cache.size,
      maxSize: this.maxSize,
      hits: this.hits,
      misses: this.misses,
      hitRate: Math.round(hitRate * 100) / 100,
      ttlSeconds: this.ttlSeconds,
    };
  }

  /**
   * Clear all cache entries.
   * v6.36.0: Session-level cache management.
   */
  clear(): void {
    this.cache.clear();
    this.hits = 0;
    this.misses = 0;
  }
}

// Global cache instance (deprecated, use createQueryCache for session-level)
let globalCache: QueryCache | undefined;

/**
 * Get or create the global QueryCache singleton.
 * v6.36.0: Deprecated - use createQueryCache() for session-level cache.
 * @deprecated Use createQueryCache() for better memory management
 */
export function getQueryCache(maxSize: number = 1000, ttlSeconds: number = 300): QueryCache {
  if (!globalCache) {
    globalCache = new QueryCache(maxSize, ttlSeconds);
  }
  return globalCache;
}

/**
 * Create a new session-level QueryCache instance.
 * v6.36.0: Preferred way to create cache - avoids global singleton memory leak.
 */
export function createQueryCache(maxSize: number = 1000, ttlSeconds: number = 300): QueryCache {
  return new QueryCache(maxSize, ttlSeconds);
}

/**
 * Reset the global query cache instance (for testing).
 */
export function resetQueryCache(): void {
  globalCache = undefined;
}

/**
 * Clear the global query cache.
 * v6.36.0: Memory leak prevention - call on session end.
 */
export function clearGlobalQueryCache(): void {
  if (globalCache) {
    globalCache.clear();
  }
}
